### ISSUE TITLE
`prefer-const` should optionally ignore destructuring assignments

### ISSUE THREAD

**Issue by moretti** [eslint/eslint#3223]
The following block will throw a warning/error about `path` not being modified when [`prefer-const`](http://eslint.org/docs/rules/prefer-const.html) is enabled:

```.js
let { host, path } = url.parse(link);
if (host.match(/^www./)) {
 host = // ...
}
```
    `path` is never modified, use `const` instead

I know that it could be rewritten to:

```.js
const values = url.parse(link);
const { path } = values;
let { host } = values;
if (host.match(/^www./)) {
 host = // ...
}
```

but It seems like an overkill. 

I think there should probably be an additional option for disabling the rule when there is a destructuring assignment.
Any thoughts?

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by michaelficarra** [comment]
The second program clearly expresses your intent better. I guess an option would be okay, but I would highly discourage its use. The rationale would basically be laziness.

**Comment by nzakas** [comment]
I'm not sure I understand why destructuring should be considered an exception, especially when one or more of the variables doesn't change. Can you explain more about why you think this option is necessary?

**Comment by moretti** [comment]
I think it should be a less strict option, just like `no-unsued-vars` has an option for disabling the check on function arguments, for example.
`prefer-cost` is excellent when you can simply replace a `let` with a `const` statement, but this cannot always be done in conjunction with a destructuring assignment.
I think the first example above reads better than the latter, but this is just a matter of my personal style (and definitely not laziness).
I understand that features cost time and effort and have to be justified based on some sort of benefits, so feel free to close the issue if you think this feature it's not necessary.

**Comment by michaelficarra** [comment]
One of the declared bindings *is* constant, and you're choosing not to tell future readers of your code. This rule asks that you always let them know this. I don't think that's unreasonable. Still :-1: from me. You can disable this rule for those individual declarations you want skipped.

**Comment by moretti** [comment]
This rule already does something similar in the initialization expression of for loops.

From the [examples](http://eslint.org/docs/rules/prefer-const.html):

```.js
for (let i = 0, end = 10; i < end; ++i) { // `end` is never modified, but we cannot separate the declarations without modifying the scope.
    console.log(a);
}
```

Just curious, why is this considered reasonable?

**Comment by moretti** [comment]
cc to @mysticatea, who [implemented](https://github.com/eslint/eslint/pull/2639) the rule (and a big thanks for that :smile:)

**Comment by nzakas** [comment]
@moretti you cannot use `const` in a loop, it causes an error.

My question to you: given that you can manually disable the rule wherever you want via `/*eslint-disable prefer-const*/`, do you still feel the need for an option?

**Comment by moretti** [comment]
@nzakas That's what I'm currently doing.
But I do not understand why this code is acceptable:

```.js
for (let i = 0, end = 10; i < end; ++i) {
   // ...
```
when it could be written as

```.js
const end = 10;
for (let i = 0; i < end; ++i) {
   // ...
```

or simply:

```.js
for (let i = 0; i < 10; ++i) {
   // ...
```

and mine is not.

**Comment by michaelficarra** [comment]
Oh nice, that looks like a bug! Please file it separately.

**Comment by moretti** [comment]
@michaelficarra  It's not a bug, it's by design. 
Please check out the examples in the official documentation:
http://eslint.org/docs/rules/prefer-const.html

**Comment by michaelficarra** [comment]
It is only by design if it declares a single binding. This rule should report when a variable declaration in that position has multiple declarators or a destructuring with multiple bindings.

**Comment by mysticatea** [comment]
```js
for (let i = 0, end = xs.length; i < end; ++i) {
    // ...
}
```

This is `let` version of one of popular patterns for performance (although needless in V8/JSC).
I made the rule not warning this `end` variable, because this `end` is inside of the loop scope and I cannot make this `end` to be `const` *without changing the scope*. It's a matter of preference: small scoping v.s. const.

In a case of destructuring, I can rewrite the code with `const`, so I'm not sure whether or not the option is needed. The option might be good, but I think I will not use it...

I would write such as:

```js
const {host: inputHost, path} = url.parse(link);
const host = /^www\./.test(inputHost) ? ... : ...; // or normalizeHost(inputHost)
```

Or

```js
function parseLink(link) {
    const {host, path} = url.parse(link);
    if (/^www\./.test(host)) {
        //...
        return {host: ..., path};
    }
    return {host, path};
}
```

**Comment by michaelficarra** [comment]
> I cannot make this `end` to be `const` *without changing the scope*

Ah, very good point, @mysticatea.

**Comment by moretti** [comment]
Thank you for your clarification, @mysticatea :smile: 
I'll close the issue because, like you're saying, it's easier to rewrite the code in a way that avoids the problem.
